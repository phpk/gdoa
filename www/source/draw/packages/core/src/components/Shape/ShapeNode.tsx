import { observer } from 'mobx-react-lite'
import * as React from 'react'
import type { IShapeTreeNode, TLShape } from '~types'
import { Shape } from './Shape'
import type { TLShapeUtilsMap } from '~TLShapeUtil'

interface ShapeNodeProps<T extends TLShape> extends IShapeTreeNode<T> {
  utils: TLShapeUtilsMap<TLShape>
}

export const ShapeNode = observer(function ShapeNode<T extends TLShape>({
  shape,
  utils,
  meta,
  children,
  ...rest
}: ShapeNodeProps<T>) {
  return (
    <>
      <Shape shape={shape} utils={utils[shape.type as T['type']]} meta={meta} {...rest} />
      {children &&
        children.map((childNode) => (
          <ShapeNode key={childNode.shape.id} utils={utils} {...childNode} />
        ))}
    </>
  )
})
