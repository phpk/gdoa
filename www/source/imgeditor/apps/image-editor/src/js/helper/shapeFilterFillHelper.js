/**
 * @author NHN. FE Development Team <dl_javascript@nhn.com>
 * @fileoverview Shape resize helper
 */
import { fabric } from 'fabric';
import { forEach, map, extend } from 'tui-code-snippet';
import resizeHelper from '@/helper/shapeResizeHelper';
import { capitalizeString, flipObject, setCustomProperty, getCustomProperty } from '@/util';

const FILTER_OPTION_MAP = {
  pixelate: 'blocksize',
  blur: 'blur',
};
const POSITION_DIMENSION_MAP = {
  x: 'width',
  y: 'height',
};

const FILTER_NAME_VALUE_MAP = flipObject(FILTER_OPTION_MAP);

/**
 * Cached canvas image element for fill image
 * @type {boolean}
 * @private
 */
let cachedCanvasImageElement = null;

/**
 * Get background image of fill
 * @param {fabric.Object} shapeObj - Shape object
 * @returns {fabric.Image}
 * @private
 */
export function getFillImageFromShape(shapeObj) {
  const { patternSourceCanvas } = getCustomProperty(shapeObj, 'patternSourceCanvas');
  const [fillImage] = patternSourceCanvas.getObjects();

  return fillImage;
}

/**
 * Reset the image position in the filter type fill area.
 * @param {fabric.Object} shapeObj - Shape object
 * @private
 */
export function rePositionFilterTypeFillImage(shapeObj) {
  const { angle, flipX, flipY } = shapeObj;
  const fillImage = getFillImageFromShape(shapeObj);
  const rotatedShapeCornerDimension = getRotatedDimension(shapeObj);
  const { right, bottom } = rotatedShapeCornerDimension;
  let { width, height } = rotatedShapeCornerDimension;
  const diffLeft = (width - shapeObj.width) / 2;
  const diffTop = (height - shapeObj.height) / 2;
  const cropX = shapeObj.left - shapeObj.width / 2 - diffLeft;
  const cropY = shapeObj.top - shapeObj.height / 2 - diffTop;
  let left = width / 2 - diffLeft;
  let top = height / 2 - diffTop;
  const fillImageMaxSize = Math.max(width, height) + Math.max(diffLeft, diffTop);

  [left, top, width, height] = calculateFillImageDimensionOutsideCanvas({
    shapeObj,
    left,
    top,
    width,
    height,
    cropX,
    cropY,
    flipX,
    flipY,
    right,
    bottom,
  });

  fillImage.set({
    angle: flipX === flipY ? -angle : angle,
    left,
    top,
    width,
    height,
    cropX,
    cropY,
    flipX,
    flipY,
  });

  setCustomProperty(fillImage, { fillImageMaxSize });
}

/**
 * Make filter option from fabric image
 * @param {fabric.Image} imageObject - fabric image object
 * @returns {object}
 */
export function makeFilterOptionFromFabricImage(imageObject) {
  return map(imageObject.filters, (filter) => {
    const [key] = Object.keys(filter);

    return {
      [FILTER_NAME_VALUE_MAP[key]]: filter[key],
    };
  });
}

/**
 * Calculate fill image position and size for out of Canvas
 * @param {Object} options - options for position dimension calculate
 *   @param {fabric.Object} shapeObj - shape object
 *   @param {number} left - original left position
 *   @param {number} top - original top position
 *   @param {number} width - image width
 *   @param {number} height - image height
 *   @param {number} cropX - image cropX
 *   @param {number} cropY - image cropY
 *   @param {boolean} flipX - shape flipX
 *   @param {boolean} flipY - shape flipY
 * @returns {Object}
 */
function calculateFillImageDimensionOutsideCanvas({
  shapeObj,
  left,
  top,
  width,
  height,
  cropX,
  cropY,
  flipX,
  flipY,
  right,
  bottom,
}) {
  const overflowAreaPositionFixer = (type, outDistance, imageLeft, imageTop) =>
    calculateDistanceOverflowPart({
      type,
      outDistance,
      shapeObj,
      flipX,
      flipY,
      left: imageLeft,
      top: imageTop,
    });
  const [originalWidth, originalHeight] = [width, height];

  [left, top, width, height] = calculateDimensionLeftTopEdge(overflowAreaPositionFixer, {
    left,
    top,
    width,
    height,
    cropX,
    cropY,
  });

  [left, top, width, height] = calculateDimensionRightBottomEdge(overflowAreaPositionFixer, {
    left,
    top,
    insideCanvasRealImageWidth: width,
    insideCanvasRealImageHeight: height,
    right,
    bottom,
    cropX,
    cropY,
    originalWidth,
    originalHeight,
  });

  return [left, top, width, height];
}

/**
 * Calculate fill image position and size for for right bottom edge
 * @param {Function} overflowAreaPositionFixer - position fixer
 * @param {Object} options - options for position dimension calculate
 *   @param {fabric.Object} shapeObj - shape object
 *   @param {number} left - original left position
 *   @param {number} top - original top position
 *   @param {number} width - image width
 *   @param {number} height - image height
 *   @param {number} right - image right
 *   @param {number} bottom - image bottom
 *   @param {number} cropX - image cropX
 *   @param {number} cropY - image cropY
 *   @param {boolean} originalWidth - image original width
 *   @param {boolean} originalHeight - image original height
 * @returns {Object}
 */
function calculateDimensionRightBottomEdge(
  overflowAreaPositionFixer,
  {
    left,
    top,
    insideCanvasRealImageWidth,
    insideCanvasRealImageHeight,
    right,
    bottom,
    cropX,
    cropY,
    originalWidth,
    originalHeight,
  }
) {
  let [width, height] = [insideCanvasRealImageWidth, insideCanvasRealImageHeight];
  const { width: canvasWidth, height: canvasHeight } = cachedCanvasImageElement;

  if (right > canvasWidth && cropX > 0) {
    width = originalWidth - Math.abs(right - canvasWidth);
  }
  if (bottom > canvasHeight && cropY > 0) {
    height = originalHeight - Math.abs(bottom - canvasHeight);
  }

  const diff = {
    x: (insideCanvasRealImageWidth - width) / 2,
    y: (insideCanvasRealImageHeight - height) / 2,
  };

  forEach(['x', 'y'], (type) => {
    const cropDistance2 = diff[type];
    if (cropDistance2 > 0) {
      [left, top] = overflowAreaPositionFixer(type, cropDistance2, left, top);
    }
  });

  return [left, top, width, height];
}

/**
 * Calculate fill image position and size for for left top
 * @param {Function} overflowAreaPositionFixer - position fixer
 * @param {Object} options - options for position dimension calculate
 *   @param {fabric.Object} shapeObj - shape object
 *   @param {number} left - original left position
 *   @param {number} top - original top position
 *   @param {number} width - image width
 *   @param {number} height - image height
 *   @param {number} cropX - image cropX
 *   @param {number} cropY - image cropY
 * @returns {Object}
 */
function calculateDimensionLeftTopEdge(
  overflowAreaPositionFixer,
  { left, top, width, height, cropX, cropY }
) {
  const dimension = {
    width,
    height,
  };

  forEach(['x', 'y'], (type) => {
    const cropDistance = type === 'x' ? cropX : cropY;
    const compareSize = dimension[POSITION_DIMENSION_MAP[type]];
    const standardSize = cachedCanvasImageElement[POSITION_DIMENSION_MAP[type]];

    if (compareSize > standardSize) {
      const outDistance = (compareSize - standardSize) / 2;

      dimension[POSITION_DIMENSION_MAP[type]] = standardSize;
      [left, top] = overflowAreaPositionFixer(type, outDistance, left, top);
    }
    if (cropDistance < 0) {
      [left, top] = overflowAreaPositionFixer(type, cropDistance, left, top);
    }
  });

  return [left, top, dimension.width, dimension.height];
}

/**
 * Make fill property of dynamic pattern type
 * @param {fabric.Image} canvasImage - canvas background image
 * @param {Array} filterOption - filter option
 * @param {fabric.StaticCanvas} patternSourceCanvas - fabric static canvas
 * @returns {Object}
 */
export function makeFillPatternForFilter(canvasImage, filterOption, patternSourceCanvas) {
  const copiedCanvasElement = getCachedCanvasImageElement(canvasImage);
  const fillImage = makeFillImage(copiedCanvasElement, canvasImage.angle, filterOption);
  patternSourceCanvas.add(fillImage);

  const fabricProperty = {
    fill: new fabric.Pattern({
      source: patternSourceCanvas.getElement(),
      repeat: 'no-repeat',
    }),
  };

  setCustomProperty(fabricProperty, { patternSourceCanvas });

  return fabricProperty;
}

/**
 * Reset fill pattern canvas
 * @param {fabric.StaticCanvas} patternSourceCanvas - fabric static canvas
 */
export function resetFillPatternCanvas(patternSourceCanvas) {
  const [innerImage] = patternSourceCanvas.getObjects();
  let { fillImageMaxSize } = getCustomProperty(innerImage, 'fillImageMaxSize');
  fillImageMaxSize = Math.max(1, fillImageMaxSize);

  patternSourceCanvas.setDimensions({
    width: fillImageMaxSize,
    height: fillImageMaxSize,
  });
  patternSourceCanvas.renderAll();
}

/**
 * Remake filter pattern image source
 * @param {fabric.Object} shapeObj - Shape object
 * @param {fabric.Image} canvasImage - canvas background image
 */
export function reMakePatternImageSource(shapeObj, canvasImage) {
  const { patternSourceCanvas } = getCustomProperty(shapeObj, 'patternSourceCanvas');
  const [fillImage] = patternSourceCanvas.getObjects();
  const filterOption = makeFilterOptionFromFabricImage(fillImage);

  patternSourceCanvas.remove(fillImage);

  const copiedCanvasElement = getCachedCanvasImageElement(canvasImage, true);
  const newFillImage = makeFillImage(copiedCanvasElement, canvasImage.angle, filterOption);

  patternSourceCanvas.add(newFillImage);
}

/**
 * Calculate a point line outside the canvas.
 * @param {fabric.Image} canvasImage - canvas background image
 * @param {boolean} reset - default is false
 * @returns {HTMLImageElement}
 */
export function getCachedCanvasImageElement(canvasImage, reset = false) {
  if (!cachedCanvasImageElement || reset) {
    cachedCanvasImageElement = canvasImage.toCanvasElement();
  }

  return cachedCanvasImageElement;
}

/**
 * Calculate fill image position for out of Canvas
 * @param {string} type - 'x' or 'y'
 * @param {fabric.Object} shapeObj - shape object
 * @param {number} outDistance - distance away
 * @param {number} left - original left position
 * @param {number} top - original top position
 * @returns {Array}
 */
function calculateDistanceOverflowPart({ type, shapeObj, outDistance, left, top, flipX, flipY }) {
  const shapePointNavigation = getShapeEdgePoint(shapeObj);
  const shapeNeighborPointNavigation = [
    [1, 2],
    [0, 3],
    [0, 3],
    [1, 2],
  ];
  const linePointsOutsideCanvas = calculateLinePointsOutsideCanvas(
    type,
    shapePointNavigation,
    shapeNeighborPointNavigation
  );
  const reatAngles = calculateLineAngleOfOutsideCanvas(
    type,
    shapePointNavigation,
    linePointsOutsideCanvas
  );
  const { startPointIndex } = linePointsOutsideCanvas;
  const diffPosition = getReversePositionForFlip({
    outDistance,
    startPointIndex,
    flipX,
    flipY,
    reatAngles,
  });

  return [left + diffPosition.left, top + diffPosition.top];
}

/**
 * Calculate fill image position for out of Canvas
 * @param {number} outDistance - distance away
 * @param {boolean} flipX - flip x statux
 * @param {boolean} flipY - flip y statux
 * @param {Array} reatAngles - Line angle of the rectangle vertex.
 * @returns {Object} diffPosition
 */
function getReversePositionForFlip({ outDistance, startPointIndex, flipX, flipY, reatAngles }) {
  const rotationChangePoint1 = outDistance * Math.cos((reatAngles[0] * Math.PI) / 180);
  const rotationChangePoint2 = outDistance * Math.cos((reatAngles[1] * Math.PI) / 180);
  const isForward = startPointIndex === 2 || startPointIndex === 3;
  const diffPosition = {
    top: isForward ? rotationChangePoint1 : rotationChangePoint2,
    left: isForward ? rotationChangePoint2 : rotationChangePoint1,
  };

  if (isReverseLeftPositionForFlip(startPointIndex, flipX, flipY)) {
    diffPosition.left = diffPosition.left * -1;
  }
  if (isReverseTopPositionForFlip(startPointIndex, flipX, flipY)) {
    diffPosition.top = diffPosition.top * -1;
  }

  return diffPosition;
}

/**
 * Calculate a point line outside the canvas.
 * @param {string} type - 'x' or 'y'
 * @param {Array} shapePointNavigation - shape edge positions
 *   @param {Object} shapePointNavigation.lefttop - left top position
 *   @param {Object} shapePointNavigation.righttop - right top position
 *   @param {Object} shapePointNavigation.leftbottom - lefttop position
 *   @param {Object} shapePointNavigation.rightbottom - rightbottom position
 * @param {Array} shapeNeighborPointNavigation - Array to find adjacent edges.
 * @returns {Object}
 */
function calculateLinePointsOutsideCanvas(
  type,
  shapePointNavigation,
  shapeNeighborPointNavigation
) {
  let minimumPoint = 0;
  let minimumPointIndex = 0;
  forEach(shapePointNavigation, (point, index) => {
    if (point[type] < minimumPoint) {
      minimumPoint = point[type];
      minimumPointIndex = index;
    }
  });

  const [endPointIndex1, endPointIndex2] = shapeNeighborPointNavigation[minimumPointIndex];

  return {
    startPointIndex: minimumPointIndex,
    endPointIndex1,
    endPointIndex2,
  };
}

/**
 * Calculate a point line outside the canvas.
 * @param {string} type - 'x' or 'y'
 * @param {Array} shapePointNavigation - shape edge positions
 *   @param {object} shapePointNavigation.lefttop - left top position
 *   @param {object} shapePointNavigation.righttop - right top position
 *   @param {object} shapePointNavigation.leftbottom - lefttop position
 *   @param {object} shapePointNavigation.rightbottom - rightbottom position
 * @param {Object} linePointsOfOneVertexIndex - Line point of one vertex
 *   @param {Object} linePointsOfOneVertexIndex.startPoint - start point index
 *   @param {Object} linePointsOfOneVertexIndex.endPointIndex1 - end point index
 *   @param {Object} linePointsOfOneVertexIndex.endPointIndex2 - end point index
 * @returns {Object}
 */
function calculateLineAngleOfOutsideCanvas(type, shapePointNavigation, linePointsOfOneVertexIndex) {
  const { startPointIndex, endPointIndex1, endPointIndex2 } = linePointsOfOneVertexIndex;
  const horizontalVerticalAngle = type === 'x' ? 180 : 270;

  return map([endPointIndex1, endPointIndex2], (pointIndex) => {
    const startPoint = shapePointNavigation[startPointIndex];
    const endPoint = shapePointNavigation[pointIndex];
    const diffY = startPoint.y - endPoint.y;
    const diffX = startPoint.x - endPoint.x;

    return (Math.atan2(diffY, diffX) * 180) / Math.PI - horizontalVerticalAngle;
  });
}

/* eslint-disable complexity */
/**
 * Calculate a point line outside the canvas for horizontal.
 * @param {number} startPointIndex - start point index
 * @param {boolean} flipX - flip x statux
 * @param {boolean} flipY - flip y statux
 * @returns {boolean} flipY - flip y statux
 */
function isReverseLeftPositionForFlip(startPointIndex, flipX, flipY) {
  return (
    (((!flipX && flipY) || (!flipX && !flipY)) && startPointIndex === 0) ||
    (((flipX && flipY) || (flipX && !flipY)) && startPointIndex === 1) ||
    (((!flipX && !flipY) || (!flipX && flipY)) && startPointIndex === 2) ||
    (((flipX && !flipY) || (flipX && flipY)) && startPointIndex === 3)
  );
}
/* eslint-enable complexity */

/* eslint-disable complexity */
/**
 * Calculate a point line outside the canvas for vertical.
 * @param {number} startPointIndex - start point index
 * @param {boolean} flipX - flip x statux
 * @param {boolean} flipY - flip y statux
 * @returns {boolean} flipY - flip y statux
 */
function isReverseTopPositionForFlip(startPointIndex, flipX, flipY) {
  return (
    (((flipX && !flipY) || (!flipX && !flipY)) && startPointIndex === 0) ||
    (((!flipX && !flipY) || (flipX && !flipY)) && startPointIndex === 1) ||
    (((flipX && flipY) || (!flipX && flipY)) && startPointIndex === 2) ||
    (((!flipX && flipY) || (flipX && flipY)) && startPointIndex === 3)
  );
}
/* eslint-enable complexity */

/**
 * Shape edge points
 * @param {fabric.Object} shapeObj - Selected shape object on canvas
 * @returns {Array} shapeEdgePoint - shape edge positions
 */
function getShapeEdgePoint(shapeObj) {
  return [
    shapeObj.getPointByOrigin('left', 'top'),
    shapeObj.getPointByOrigin('right', 'top'),
    shapeObj.getPointByOrigin('left', 'bottom'),
    shapeObj.getPointByOrigin('right', 'bottom'),
  ];
}

/**
 * Rotated shape dimension
 * @param {fabric.Object} shapeObj - Shape object
 * @returns {Object} Rotated shape dimension
 */
function getRotatedDimension(shapeObj) {
  const [{ x: ax, y: ay }, { x: bx, y: by }, { x: cx, y: cy }, { x: dx, y: dy }] =
    getShapeEdgePoint(shapeObj);

  const left = Math.min(ax, bx, cx, dx);
  const top = Math.min(ay, by, cy, dy);
  const right = Math.max(ax, bx, cx, dx);
  const bottom = Math.max(ay, by, cy, dy);

  return {
    left,
    top,
    right,
    bottom,
    width: right - left,
    height: bottom - top,
  };
}

/**
 * Make fill image
 * @param {HTMLImageElement} copiedCanvasElement - html image element
 * @param {number} currentCanvasImageAngle - current canvas angle
 * @param {Array} filterOption - filter option
 * @returns {fabric.Image}
 * @private
 */
function makeFillImage(copiedCanvasElement, currentCanvasImageAngle, filterOption) {
  const fillImage = new fabric.Image(copiedCanvasElement);

  forEach(extend({}, ...filterOption), (value, key) => {
    const fabricFilterClassName = capitalizeString(key);
    const filter = new fabric.Image.filters[fabricFilterClassName]({
      [FILTER_OPTION_MAP[key]]: value,
    });
    fillImage.filters.push(filter);
  });
  fillImage.applyFilters();

  setCustomProperty(fillImage, {
    originalAngle: currentCanvasImageAngle,
    fillImageMaxSize: Math.max(fillImage.width, fillImage.height),
  });
  resizeHelper.adjustOriginToCenter(fillImage);

  return fillImage;
}
