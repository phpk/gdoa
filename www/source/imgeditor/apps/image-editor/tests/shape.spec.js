import { fabric } from 'fabric';
import Graphics from '@/graphics';
import Shape from '@/component/shape';
import { resize } from '@/helper/shapeResizeHelper';
import { getFillImageFromShape, getCachedCanvasImageElement } from '@/helper/shapeFilterFillHelper';

describe('Shape', () => {
  let canvas, graphics, mockImage, fEvent, shape, shapeObj;

  beforeAll(() => {
    graphics = new Graphics(document.createElement('canvas'));
    canvas = graphics.getCanvas();
    shape = new Shape(graphics);
  });

  beforeEach(() => {
    mockImage = new fabric.Image();
    graphics.setCanvasImage('mockImage', mockImage);
    fEvent = { e: {} };
  });

  afterEach(() => {
    canvas.forEachObject((obj) => {
      canvas.remove(obj);
    });
  });

  it('should be calculated correctly.', () => {
    const pointer = canvas.getPointer(fEvent.e);
    const settings = {
      strokeWidth: 0,
      type: 'rect',
      left: 150,
      top: 200,
      width: 40,
      height: 40,
      originX: 'center',
      originY: 'center',
    };
    shape.add('rect', settings);
    [shapeObj] = canvas.getObjects();
    const setSpy = jest.spyOn(shapeObj, 'set');

    resize(shapeObj, pointer);

    expect(setSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        left: settings.left - settings.width / 2,
        top: settings.top - settings.height / 2,
      })
    );
  });

  it('should be created on canvas(rect)', () => {
    shape.add('rect');
    [shapeObj] = canvas.getObjects();

    expect(shapeObj.get('type')).toBe('rect');
  });

  it('should be created on canvas(circle)', () => {
    shape.add('circle');
    [shapeObj] = canvas.getObjects();

    expect(shapeObj.get('type')).toBe('circle');
  });

  it('should be created on canvas(triangle)', () => {
    shape.add('triangle');

    [shapeObj] = canvas.getObjects();

    expect(shapeObj.type).toBe('triangle');
  });

  it('should be set the rectangle object when add() is called with no options', () => {
    shape.add('rect');
    [shapeObj] = canvas.getObjects();

    expect(shapeObj).toMatchObject({ width: 1, height: 1 }); // strokeWidth: 1, width: 1, height: 1
  });

  it('should be set the circle object when add() is called with no options', () => {
    shape.add('circle');
    [shapeObj] = canvas.getObjects();

    expect(shapeObj).toMatchObject({ width: 0, height: 0 });
  });

  it('should be set the triangle object when add() is called with no options', () => {
    shape.add('triangle');
    [shapeObj] = canvas.getObjects();

    expect(shapeObj).toMatchObject({ width: 1, height: 1 }); // strokeWidth: 1, width: 1, height: 1
  });

  it('should be set the rectangle object when add() is called with the options', () => {
    const settings = {
      fill: 'blue',
      stroke: 'red',
      strokeWidth: 10,
      type: 'rect',
      width: 100,
      height: 100,
    };
    shape.add('rect', settings);
    [shapeObj] = canvas.getObjects();

    expect(shapeObj).toMatchObject(settings);
  });

  it('should be set the circle object when add() is called with the options', () => {
    const settings = {
      fill: 'blue',
      stroke: 'red',
      strokeWidth: 3,
      type: 'circle',
      rx: 100,
      ry: 50,
    };
    shape.add('circle', settings);
    [shapeObj] = canvas.getObjects();

    expect(shapeObj).toMatchObject(settings);
  });

  it('should be set the triangle object when add() is called with the options', () => {
    const settings = {
      fill: 'blue',
      stroke: 'red',
      strokeWidth: 0,
      type: 'triangle',
      width: 100,
      height: 100,
    };
    shape.add('triangle', settings);
    [shapeObj] = canvas.getObjects();

    expect(shapeObj).toMatchObject(settings);
  });

  it('should be changed when change() is called(rect)', () => {
    const settings = { fill: 'blue', stroke: 'red', width: 10, height: 20 };
    shape.add('rect');
    [shapeObj] = canvas.getObjects();
    shape.change(shapeObj, settings);

    expect(shapeObj).toMatchObject(settings);
  });

  it('should be changed when change() is called(circle)', () => {
    const settings = { fill: 'blue', stroke: 'red', rx: 10, ry: 20 };
    shape.add('circle');
    [shapeObj] = canvas.getObjects();
    shape.change(shapeObj, settings);

    expect(shapeObj).toMatchObject(settings);
  });

  it('should be changed when change() is called(triangle)', () => {
    const settings = { fill: 'blue', stroke: 'red', width: 10, height: 20 };
    shape.add('triangle');
    [shapeObj] = canvas.getObjects();
    shape.change(shapeObj, settings);

    expect(shapeObj).toMatchObject(settings);
  });

  describe('Fill - filter type', () => {
    beforeEach(() => {
      getCachedCanvasImageElement(canvas, true);
      mockImage = new fabric.Image();
      graphics.setCanvasImage('mockImage', mockImage);
      shape.add('rect', {
        strokeWidth: 0,
        left: 20,
        top: 30,
        width: 100,
        height: 80,
        fill: {
          type: 'filter',
          filter: [{ pixelate: 20 }],
        },
      });
      [shapeObj] = canvas.getObjects();
    });

    it('should be executed when a movement, rotation, and scaling event of a filter type fill is applied', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 10, y: 10 });
      const _resetPositionFillFilterSpy = jest.spyOn(shape, '_resetPositionFillFilter');
      shapeObj.fire('moving');
      shapeObj.fire('rotating');
      shapeObj.fire('scaling');

      expect(_resetPositionFillFilterSpy).toHaveBeenCalledTimes(3);
    });

    it('should be changed cropX and cropY values of the image filled with the shape background', () => {
      shape._resetPositionFillFilter(shapeObj);
      const fillImage = getFillImageFromShape(shapeObj);

      expect(fillImage).toMatchObject({ cropX: -30, cropY: -10 });
    });

    it('should be the same size as the shape', () => {
      shape._resetPositionFillFilter(shapeObj);
      const fillImage = getFillImageFromShape(shapeObj);

      expect(fillImage).toMatchObject({ width: 100, height: 80 });
    });

    it('should be the same size as the rectangle that draws the rotated object border', () => {
      shapeObj.set({ angle: 40 });
      shape._resetPositionFillFilter(shapeObj);
      const fillImage = getFillImageFromShape(shapeObj);

      expect(fillImage).toMatchSnapshot();
    });

    it('should have the shape reverse rotation value if repositioning is performed while the angle is changed', () => {
      shapeObj.set({ angle: 40 });
      shape._resetPositionFillFilter(shapeObj);
      const { angle } = getFillImageFromShape(shapeObj);

      expect(angle).toBe(-40);
    });

    it('should give the expected result for shapes that go outside the bottom right area of the canvas', async () => {
      const obj = await shape.add('rect', {
        strokeWidth: 0,
        left: 250,
        top: 100,
        width: 200,
        height: 200,
        fill: {
          type: 'filter',
          filter: [{ pixelate: 20 }],
        },
      });
      shapeObj = graphics.getObject(obj.id);
      const fillImage = getFillImageFromShape(shapeObj);

      expect(fillImage).toMatchObject({ top: 75, left: 75, width: 150, height: 150 });
    });

    it('should give the expected result for shapes that go outside the top left area of the canvas', async () => {
      const obj = await shape.add('rect', {
        strokeWidth: 0,
        left: 50,
        top: 30,
        width: 200,
        height: 70,
        fill: {
          type: 'filter',
          filter: [{ pixelate: 20 }],
        },
      });
      shapeObj = graphics.getObject(obj.id);
      const fillImage = getFillImageFromShape(shapeObj);

      expect(fillImage).toMatchSnapshot();
    });

    it('should have the applied filter', () => {
      const fillImage = getFillImageFromShape(shapeObj);

      expect(fillImage.filters).not.toHaveLength(0);
    });
  });

  describe('_onFabricMouseMove()', () => {
    beforeEach(() => {
      shape.add('rect', { left: 100, top: 100 });
      [shapeObj] = canvas.getObjects();
      shape._shapeObj = shapeObj;
    });

    it('should be set to "left" and "top" when the mouse direction is in 1th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 200, y: 120 });

      shape._onFabricMouseMove(fEvent);

      expect(shapeObj).toMatchObject({ originX: 'left', originY: 'top' });
    });

    it('should be set to "right" and "top" when the mouse direction is in 2th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 80, y: 100 });

      shape._onFabricMouseMove(fEvent);

      expect(shapeObj).toMatchObject({ originX: 'right', originY: 'top' });
    });

    it('should be set to "right" and "bottom" when the mouse direction is in 3th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 80, y: 80 });

      shape._onFabricMouseMove(fEvent);

      expect(shapeObj).toMatchObject({ originX: 'right', originY: 'bottom' });
    });

    it('should be set to "left" and "bottom" when the mouse direction is in 4th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 200, y: 80 });

      shape._onFabricMouseMove(fEvent);

      expect(shapeObj).toMatchObject({ originX: 'left', originY: 'bottom' });
    });
  });

  describe('_onFabricMouseUp()', () => {
    let startPoint;

    beforeEach(() => {
      shape.add('circle', { left: 100, top: 100 });
      [shapeObj] = canvas.getObjects();
      shape._shapeObj = shapeObj;
    });

    it('should be the same as start point when the drawing shape is in 1th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 200, y: 120 });
      startPoint = shapeObj.getPointByOrigin('left', 'top');

      shape._onFabricMouseMove(fEvent);
      shape._onFabricMouseUp();

      expect(shapeObj.getPointByOrigin('left', 'top')).toEqual(startPoint);
    });

    it('should be the same as start point when the drawing shape is in 2th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 80, y: 120 });

      startPoint = shapeObj.getPointByOrigin('right', 'top');

      shape._onFabricMouseMove(fEvent);
      shape._onFabricMouseUp();

      expect(shapeObj.getPointByOrigin('right', 'top')).toEqual(startPoint);
    });

    it('should be the same as start point when the drawing shape is in 3th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 80, y: 80 });

      startPoint = shapeObj.getPointByOrigin('right', 'bottom');

      shape._onFabricMouseMove(fEvent);
      shape._onFabricMouseUp();

      expect(shapeObj.getPointByOrigin('right', 'bottom')).toEqual(startPoint);
    });

    it('should be the same as start point when the drawing shape is in 4th quadrant', () => {
      jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 120, y: 80 });

      startPoint = shapeObj.getPointByOrigin('left', 'bottom');

      shape._onFabricMouseMove(fEvent);
      shape._onFabricMouseUp();

      expect(shapeObj.getPointByOrigin('left', 'bottom')).toEqual(startPoint);
    });
  });

  it('should have the same "width" and "height" values when drawing the shape with mouse and the "isRegular" option set to true(x-axis)', () => {
    shape.add('rect', { left: 0, top: 0 });
    shape._withShiftKey = true;
    [shapeObj] = canvas.getObjects();
    shape._shapeObj = shapeObj;
    jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 200, y: 100 });

    shape._onFabricMouseMove(fEvent);
    shape._onFabricMouseUp();

    expect(shapeObj).toMatchObject({ width: 200, height: 200 });
  });

  it('should have the same "width" and "height" values when drawing the shape with mouse and the "isRegular" option set to true(y-axis)', () => {
    shape.add('rect', { left: 0, top: 0 });
    shape._withShiftKey = true;
    [shapeObj] = canvas.getObjects();
    shape._shapeObj = shapeObj;
    jest.spyOn(canvas, 'getPointer').mockReturnValue({ x: 100, y: 200 });

    shape._onFabricMouseMove(fEvent);
    shape._onFabricMouseUp();

    expect(shapeObj).toMatchObject({ width: 200, height: 200 });
  });
});
