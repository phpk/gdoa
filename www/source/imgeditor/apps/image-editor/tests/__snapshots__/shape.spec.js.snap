// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shape Fill - filter type should be the same size as the rectangle that draws the rotated object border 1`] = `
Object {
  "angle": -40,
  "backgroundColor": "",
  "cropX": -44.01372654341047,
  "cropY": -32.78115820908609,
  "crossOrigin": null,
  "fill": "rgb(0,0,0)",
  "fillRule": "nonzero",
  "filters": Array [
    Object {
      "blocksize": 20,
      "type": "Pixelate",
    },
  ],
  "flipX": false,
  "flipY": false,
  "globalCompositeOperation": "source-over",
  "height": 125.56,
  "left": 104.79,
  "opacity": 1,
  "originX": "center",
  "originY": "center",
  "paintFirst": "fill",
  "scaleX": 1,
  "scaleY": 1,
  "shadow": null,
  "skewX": 0,
  "skewY": 0,
  "src": "data:image/png;base64,00",
  "stroke": null,
  "strokeDashArray": null,
  "strokeDashOffset": 0,
  "strokeLineCap": "butt",
  "strokeLineJoin": "miter",
  "strokeMiterLimit": 4,
  "strokeUniform": false,
  "strokeWidth": 0,
  "top": 36.82,
  "type": "image",
  "version": "4.6.0",
  "visible": true,
  "width": 128.03,
}
`;

exports[`Shape Fill - filter type should give the expected result for shapes that go outside the top left area of the canvas 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "",
  "cropX": -50,
  "cropY": -5,
  "crossOrigin": null,
  "fill": "rgb(0,0,0)",
  "fillRule": "nonzero",
  "filters": Array [
    Object {
      "blocksize": 20,
      "type": "Pixelate",
    },
  ],
  "flipX": false,
  "flipY": false,
  "globalCompositeOperation": "source-over",
  "height": 70,
  "left": 150,
  "opacity": 1,
  "originX": "center",
  "originY": "center",
  "paintFirst": "fill",
  "scaleX": 1,
  "scaleY": 1,
  "shadow": null,
  "skewX": 0,
  "skewY": 0,
  "src": "data:image/png;base64,00",
  "stroke": null,
  "strokeDashArray": null,
  "strokeDashOffset": 0,
  "strokeLineCap": "butt",
  "strokeLineJoin": "miter",
  "strokeMiterLimit": 4,
  "strokeUniform": false,
  "strokeWidth": 0,
  "top": 40,
  "type": "image",
  "version": "4.6.0",
  "visible": true,
  "width": 200,
}
`;
